### PR TITLE
Feature authorize

### PR DESCRIPTION
**Pull request by tathai2610**
# Description
- Add authorization
  - The user now has limited access to different pages.
- Add admin's approval to open shop
  - When a user creates their shop, admin will receive those requests and decide whether to approve.
- Add states to shops
  - Currently, there are 3 states: `pending`, `active` and `rejected`
- Add admin views, update UIs
- Add tests for authorization 

# Screenshots 
### Admin::Dashboard#index

Admin::Dashboard#index

### Admin::Dashboard#pending_shops
Admin::Dashboard#pending_shops

### Shop#show _(updated)_
Shop#show

### CODE REVIEW

File: app/models/shop.rb [hunk 1]
```diff
@@ -8,6 +8,25 @@ class Shop < ApplicationRecord
...
+    event :reject do 
+      transition pending: :rejected 
+    end
+
+    state :active 
+    state :pending
+    state :rejected
+  end
+
+  def initialize(state="pending")
```
Comment (nathan-phan-goldenowl) [inline app/models/shop.rb:25]: prefer to use symbol for state

File: app/policies/admin/dashboard_policy.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+  def index? 
+    user.present? && user.has_role?(:admin)
+  end
+
+  def pending_shops? 
+    index?
+  end
+
+  def handle_shop? 
+    index?
```
Comment (nathan-phan-goldenowl) [inline app/policies/admin/dashboard_policy.rb:18]: should authorize if shop is pending

File: app/controllers/admin/dashboard_controller.rb [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    authorize current_user, policy_class: Admin::DashboardPolicy
+  end
+
+  def pending_shops 
+    @pending_shops = Shop.where(state: "pending").includes(:user)
+    authorize @pending_shops, policy_class: Admin::DashboardPolicy
+  end
+
+  def handle_shop
+    @shop = Shop.find(params[:shop_id])
```
Comment (nathan-phan-goldenowl) [inline app/controllers/admin/dashboard_controller.rb:14]: Can't make sure @shop is a pending one.

Should change to 
```
Shop.pendings.find(params[:shop_id])
```

and add scope `pendings` to Shop model for re-use

File: app/controllers/shops_controller.rb [hunk 4]
```diff
@@ -33,14 +32,16 @@ def set_user
...
+    authorize @shop
   end
 
   def shop_params 
     params.require(:shop).permit(:name, :description)
   end
 
   def check_shop_exist
-    if current_user.shop.present?
+    authorize Shop.new
```
Comment (nathan-phan-goldenowl) [inline app/controllers/shops_controller.rb:43]: why do we authorize new shop here?

File: app/models/shop.rb (hunk 1 above)
Comment (tathai2610) [reply app/models/shop.rb:25]: noted and updated

File: app/policies/admin/dashboard_policy.rb (hunk 2 above)
Comment (tathai2610) [reply app/policies/admin/dashboard_policy.rb:18]: noted and updated
```
def handle_shop? 
  index? && record.pending?
end
```

File: app/controllers/admin/dashboard_controller.rb (hunk 3 above)
Comment (tathai2610) [reply app/controllers/admin/dashboard_controller.rb:14]: noted and updated

File: app/controllers/shops_controller.rb (hunk 4 above)
Comment (tathai2610) [reply app/controllers/shops_controller.rb:43]: This is my mistake, and I realize that we don't need to authorize anything there, so I delete that line.

**Pull request merged**